lib/readyset/model_extension.rb

```diff
@@ -0,0 +1,40 @@
+module Readyset
+  module ModelExtension
+    extend ActiveSupport::Concern
+
+    prepended do
+      require 'active_support'
+
+      # Defines a new class method on a model that wraps the ActiveRecord query in `body` in a call
+      # to `Readyset.route`.
+      #
+      # NOTE: `body` should consist of nothing other than an ActiveRecord query! If you need to run
+      # actions before or after the query execution, you should wrap the invocation of the named
+      # query in another method.
+      #
+      # @param [Symbol] name the name of the method that will be defined
+      # @param [Proc] body a lambda that wraps an ActiveRecord query
+      def self.readyset_query(name, body)
+        unless body.respond_to?(:call)
+          raise ArgumentError, 'The query body needs to be callable.'
+        end
+
+        if dangerous_class_method?(name)
+          raise ArgumentError, "You tried to define a ReadySet query named \"#{name}\" " \
+            "on the model \"#{self.name}\", but Active Record already defined " \
+            'a class method with the same name.'
+        end
+
+        if method_defined_within?(name, ActiveRecord::Relation)
+          raise ArgumentError, "You tried to define a ReadySet query named \"#{name}\" " \
+            "on the model \"#{self.name}\", but ActiveRecord::Relation already defined " \
+            'an instance method with the same name.'
+        end
```

Comment:
Good error handling!

Reply:
Whoops! I meant to leave a comment here: I used the definition of Rails's `scope` method as a guide here, since it similarly defines a class method that is included in `ActiveRecord::Base`: https://github.com/rails/rails/blob/0ad2050c6e6024730d6a94d1d0090bd61b699bb7/activerecord/lib/active_record/scoping/named.rb#L154-L155

But yep, this ensures that users don't overwrite methods defined by `ActiveRecord` or `ActiveRecord::Relation`